### PR TITLE
Rewrite how default widgets in the docmanager work.

### PR DESCRIPTION
**Pull request by jasongrout**
I also changed the wildcard extension from .* to just *, which seems to make more sense. It means we can check for wildcard by just looking at the first character, and wildcards apply even when there is no extension (i.e., no `.`), so including the `.` seems wrong.

I use Sets a lot, which make the data structures much simpler.

### CODE REVIEW

File: src/docmanager/manager.ts [hunk 1]
```diff
@@ -30,6 +30,26 @@ import {
...
 
+/**
+ * Return the extension, given a path.
+ *
+ * @param path - the file path.
+ *
+ * @returns the extension for the filepath, with the '.' included, or '' if no extension.
+ *
+ * #### TODO
+ * This should call the jupyter-js-services to get the extension
```
Comment (blink1073) [inline src/docmanager/manager.ts]: cf https://github.com/jupyter/jupyter-js-services/pull/194

File: src/docregistry/interfaces.ts [hunk 2]
```diff
@@ -247,7 +247,9 @@ interface IWidgetFactoryOptions {
    * The file extensions for which the factory should be the default.
    *
    * #### Notes
-   * Use ".*" to denote all files.
+   * Use "*" to denote all files. Specific file extensions must be preceded with '.', like '.png', '.txt', etc. Entries in this attribute must also be included in the fileExtensions attribute.
```
Comment (blink1073) [inline src/docregistry/interfaces.ts]: Break the comment for line wrap

File: src/docregistry/registry.ts [hunk 3]
```diff
@@ -208,47 +213,57 @@ class DocumentRegistry implements IDisposable {
...
     }
-    // Add the default widget if it was not already added.
-    name = this._defaultWidgetFactory;
-    if (name && factories.indexOf(name) === -1) {
-      options = this._widgetFactories[name];
-      if (options.modelName in this._modelFactories) {
-        factories.push(name);
-      }
+
+    // Add the gobal default factory
```
Comment (blink1073) [inline src/docregistry/registry.ts]: global*

Comment (blink1073) [inline src/docregistry/registry.ts]: Missing periods on this and the next comment.

**Comment by blink1073** [comment]
Looks good modulo style comments.

File: src/docregistry/registry.ts (hunk 3 above)
Comment (jasongrout) [inline src/docregistry/registry.ts]: I didn't realize we were making comments complete sentences.

File: src/docmanager/manager.ts (hunk 1 above)
Comment (jasongrout) [inline src/docmanager/manager.ts]: Yep, when it goes in I'll put in a PR to use it instead.

**Comment by jasongrout** [comment]
Fixed style issues.

**Comment by blink1073** [comment]
Thank ye!

**Pull request merged**